h/traversal/group_membership.py

```diff
@@ -0,0 +1,51 @@
+from dataclasses import dataclass
+
+from pyramid.httpexceptions import HTTPNotFound
+
+from h.exceptions import InvalidUserId
+from h.models import Group, GroupMembership, User
+
+
+@dataclass
+class GroupMembershipContext:
+    group: Group
+    user: User
+    membership: GroupMembership | None
+
+
+def group_membership_api_factory(request) -> GroupMembershipContext:
+    user_service = request.find_service(name="user")
+    group_service = request.find_service(name="group")
+    group_members_service = request.find_service(name="group_members")
+
+    userid = request.matchdict["userid"]
+    pubid = request.matchdict["pubid"]
+
+    def get_user() -> User | None:
+        if userid == "me":
+            if request.authenticated_userid:
+                return user_service.fetch(request.authenticated_userid)
+
+            return None
+
+        try:
+            return user_service.fetch(userid)
+        except InvalidUserId:
+            return None
+
+    user = get_user()
+
+    if not user:
+        raise HTTPNotFound(f"User not found: {userid}")
+
+    group = group_service.fetch(pubid)
+
+    if not group:
+        raise HTTPNotFound(f"Group not found: {pubid}")
+
+    membership = group_members_service.get(group, user)
+
+    if not membership and request.method != "POST":
+        raise HTTPNotFound(f"Membership not found: ({pubid}, {userid})")
```

Comment:
Special case for `POST /api/groups/{pubid}/members/{userid}`: this request is _creating_ a new membership, so it doesn't expect the membership to already exist. An alternative implementation would be for `POST` requests to be routed to a different context factory (possibly even returning a different context object).

Note that this doesn't guarantee that the membership doesn't already exist for `POST` requests: the view code itself has to handle the case of trying to create a membership that doesn't exist. For example the view could raise an error. Or it could make this case a no-op. Alternatively we _could_ handle that here and raise some appropriate HTTP exception when trying to `POST` a membership that already exists. The existing `add_member()` view already handles this case and it's not the view that I'm interested in in this PR, so for now just leaving it for the existing view to continue handling as it has already done.